### PR TITLE
Cleanup/fix context package

### PR DESCRIPTION
**Pull request by Nevay**
Removed:
`Context::withValue()`: Semi-broken due to creation of hidden scope
`Context::detach()`: Same functionality as `ScopeInterface::detach()`
`Context::attach()`: Same functionality as `Context::activate()`

`Context::getValue()`: Can be removed? Users will use class-specific `::fromContext()` / `::getCurrrent()` methods instead of this method anyways?

Changes:
`Context` and `ContextKey`: Changed to `final` to reduce scope / allow addition of new methods in the future without breaking backwards compatibility.
`ArrayAccessGetterSetter`: Resolves issues with case-insensitive keys and preserves casing as recommended by [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/94c9c75c4f82fbda08bddff02ce9a0c582fdd55c/specification/context/api-propagators.md#set).

### CODE REVIEW

File: src/Context/Propagation/TextMapPropagator.php [hunk 1]
```diff
@@ -14,7 +15,9 @@ public static function composite(TextMapPropagatorInterface ...$propagators): Te
             case 0:
                 return NoopTextMapPropagator::getInstance();
             case 1:
-                return $propagators[0];
+                assert($propagators !== []);
```
Comment (brettmc) [inline src/Context/Propagation/TextMapPropagator.php:18]: is this possible, given the `count()` above?

File: tests/Unit/Context/ContextTest.php [hunk 2]
```diff
@@ -146,19 +125,19 @@ public function test_reusing_key_overwrites_value(): void
 
     public function test_ctx_value_not_found_throws(): void
```
Comment (brettmc) [inline tests/Unit/Context/ContextTest.php:126]: I don't think this test's name is accurate anymore. `s/throws/returns_null/` ?

File: src/Context/Propagation/TextMapPropagator.php (hunk 1 above)
Comment (Nevay) [reply src/Context/Propagation/TextMapPropagator.php:18]: No, added to make [psalm happy](https://github.com/open-telemetry/opentelemetry-php/runs/6907417904?check_suite_focus=true#step:11:18).

Comment (brettmc) [reply src/Context/Propagation/TextMapPropagator.php:18]: ok. you could also `@psalm-suppress` which might be clearer for future devs...?

Comment (Nevay) [reply src/Context/Propagation/TextMapPropagator.php]: Seems like I accidentially used `/*` instead of `/**` when I tried that yesterday - changed.

**Review by brettmc — APPROVED** [review]

**Review by bobstrecansky — APPROVED** [review]

**Pull request merged**